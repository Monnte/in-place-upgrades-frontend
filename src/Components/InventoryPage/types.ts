export declare type SystemColumn = {
  key: string;
  sortKey?: string;
  props?: {
    width: number;
  };
  title: string;
  renderFunc?: (value, id: string) => JSX.Element;
};

export declare type SystemFilters = {
  hostnameOrId?: string;
  osFilter?: string[];
};
