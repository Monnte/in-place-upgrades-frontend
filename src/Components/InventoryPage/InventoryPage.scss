.name-column {
    font-size: var(--pf-global--FontSize--lg);
}
