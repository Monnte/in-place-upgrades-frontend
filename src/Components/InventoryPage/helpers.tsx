import React from 'react';
import { entitiesReducer } from '../../store/index';
import { SystemColumn, SystemFilters } from './types';
import ReducerRegistry from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { Reducer } from 'redux';

const buildSortString = (orderBy: string, orderDirection: string): string => {
  const sortString = orderBy ? '&sort=' : '';
  let direction = '';
  if (orderDirection === 'DESC') {
    direction = '-';
  }
  const order = orderBy === 'updated' ? 'last_seen' : orderBy;

  return `${sortString}${direction}${order}`;
};

const buildFilterString = (filters: SystemFilters): string => {
  const displayNameFilter = filters.hostnameOrId ? `&display_name=${filters.hostnameOrId}` : '';

  const osFilter = filters.osFilter?.length ? '&os_version=' + filters.osFilter.join(',') : '';

  return `${displayNameFilter}${osFilter}`;
};

export const buildFilterSortString = (limit: number, offset: number, orderBy: string, orderDirection: string, filters: SystemFilters) => {
  const limitOffsetString = `limit=${limit}&offset=${offset}`;
  const sortString = buildSortString(orderBy, orderDirection);
  const filterString = buildFilterString(filters);
  return `?${limitOffsetString}${sortString}${filterString}`;
};

export const findCheckedValue = (total: number, selected: number): boolean | null => {
  if (selected === total && total > 0) {
    return true;
  } else if (selected > 0 && selected < total) {
    return null;
  } else {
    return false;
  }
};

const createSystemLink = (id: string, name: string, keyData: string, isBetaEnv: boolean): JSX.Element => (
  <a rel="noreferrer" target="_blank" key={keyData} href={isBetaEnv ? `/preview/insights/inventory/${id}` : `/insights/inventory/${id}`}>
    {name}
  </a>
);

export const systemColumns = (isBeta: boolean): SystemColumn[] => [
  {
    key: 'display_name',
    sortKey: 'display_name',
    props: { width: 20 },
    title: 'Name',
    renderFunc: (name, id) => {
      return createSystemLink(id, name, `system-name-${id}`, isBeta);
    },
  },
  {
    key: 'tags',
    title: 'Tags',
  },
  {
    key: 'os_version',
    sortKey: 'os_version',
    props: { width: 10 },
    title: 'OS',
  },
  {
    key: 'updated',
    title: 'updated',
  },
];

export const defaultOnLoad = (columns, getRegistry: () => ReducerRegistry<Reducer>) => {
  return ({ INVENTORY_ACTION_TYPES, mergeWithEntities }) =>
    getRegistry().register({
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-ignore
      ...mergeWithEntities(entitiesReducer(INVENTORY_ACTION_TYPES, columns), {
        page: 1,
        perPage: 10,
        sortBy: {
          key: 'updated',
          direction: 'desc',
        },
      }),
    });
};
