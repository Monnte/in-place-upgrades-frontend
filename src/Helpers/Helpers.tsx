import React from 'react';
import { Text, TextContent } from '@patternfly/react-core';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/redux';

export const dispatchNotification = (store, notification) => {
  store.dispatch(addNotification(notification));
};

export const tasksExecuteSucessNotif = (store, title, ids, task_id) => {
  dispatchNotification(store, {
    variant: 'info',
    title: 'Task running',
    description: (
      <span>
        Your task &quot;{title}&quot; is running on {ids.length} system
        {ids.length > 1 ? 's' : ''}.
        <br />
        <br />
        <TextContent>
          <Text component="a" href={`/insights/tasks/executed/${task_id}`} rel="noreferrer" target="_blank">
            View progress
          </Text>
        </TextContent>
      </span>
    ),
    dismissable: true,
  });
};

export const tasksExecuteErrorNotif = (store, message) => {
  dispatchNotification(store, {
    variant: 'danger',
    title: 'Error',
    description: message,
    dismissable: true,
    autoDismiss: false,
  });
};
