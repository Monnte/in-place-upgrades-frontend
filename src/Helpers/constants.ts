export const PERMISSIONS = {
  inventoryHostsRead: ['inventory:hosts:read'],
  tasks: ['tasks:*:*'],
  useRecommendations: ['advisor:*:*'],
};

export const SERVICES = {
  inventory: 'inventory',
  tasks: 'tasks',
  advisor: 'advisor',
};

export const INVENTORY_API_ROOT = '/api/inventory/v1';
export const INVENTORY_HOSTS_ROOT = '/hosts';
export const INVENTORY_TAGS_ROOT = '/tags';

export const TASKS_API_ROOT = '/api/tasks/v1';
export const TASKS_AVAILABLE_ROOT = '/task';
export const TASK_EXECUTED_ROOT = '/executed_task';

export const RECOMMENDATIONS_API_ROOT = '/api/insights/v1';
export const RECOMMENDATIONS_RULES_ROOT = '/rule';
export const RECOMMENDATIONS_DETAIL_ROOT = '/insights/advisor/recommendations';

export const INFO_ALERT_SYSTEMS = 'Eligible systems include systems connected to console.redhat.com with rhc, or Satellite with Cloud Connector.';
