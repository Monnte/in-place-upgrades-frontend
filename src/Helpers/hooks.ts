import { useEffect, useState } from 'react';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';

const matchPermissions = (permissionA, permissionB) => {
  const segmentsA = permissionA.split(':');
  const segmentsB = permissionB.split(':');

  if (segmentsA.length !== segmentsB.length) {
    return false;
  }

  return segmentsA.every((segmentA, index) => segmentA === segmentsB[index] || segmentA === '*' || segmentsB[index] === '*');
};

export const useRbac = (requestedPermissions, app = 'upgrades') => {
  const [allPermissions, setAllPermissions] = useState([] as { permission: string }[]);
  const [loading, setLoading] = useState(true);
  const chrome = useChrome();

  useEffect(() => {
    chrome?.getUserPermissions?.(app, true).then((permissions) => {
      setAllPermissions(permissions);
      setLoading(false);
    });
  }, []);

  if (loading) {
    return [requestedPermissions.map(() => []), true];
  }

  return [
    requestedPermissions.map((requestedPermission) => allPermissions?.some((item) => matchPermissions(item.permission, requestedPermission))),
    false,
  ];
};
