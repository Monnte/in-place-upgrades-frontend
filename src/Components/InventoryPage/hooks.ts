import { inventoryFetchSystems, inventoryFetchSystemsTags } from '../../api';
import { buildAditionalFields, buildFilterString, buildPerPagePageString, buildSortString } from './Helpers';

export const useGetEntities = (onComplete: (result, filtersString) => void, { selectedIds }: { selectedIds?: string[] } = {}) => {
  return async (_items, config) => {
    const { page, per_page: perPage, orderBy, orderDirection, filters } = config;

    const sortString = buildSortString(orderBy, orderDirection);
    const filtersString = buildFilterString(filters);
    const perPageString = buildPerPagePageString(perPage, page);
    const aditionalFieldsString = buildAditionalFields(['operating_system', 'system_update_method']);
    const finalFilterSortString = `?${perPageString}${sortString}${filtersString}${aditionalFieldsString}`;

    const fetchedEntities = await inventoryFetchSystems(finalFilterSortString);
    const ids = fetchedEntities?.results?.map((entity) => entity.id) || [];
    const fetchedTags = ids.length > 0 ? await inventoryFetchSystemsTags(ids, `?per_page=${perPage}${sortString}`) : null;

    const { results, total } = fetchedEntities || {};

    onComplete && onComplete(fetchedEntities, filtersString);

    return {
      results:
        results?.map((entity) => ({
          ...entity,
          selected: (selectedIds || []).map((id) => id).includes(entity.id),
          tags: fetchedTags?.results[entity.id] || [],
        })) || [],
      page,
      perPage,
      orderBy,
      orderDirection,
      total: total,
    };
  };
};
