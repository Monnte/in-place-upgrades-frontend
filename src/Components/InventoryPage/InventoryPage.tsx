import React, { useContext, useEffect, useRef, useState } from 'react';
import propTypes from 'prop-types';
import { Bullseye, Spinner } from '@patternfly/react-core';
import { useGetEntities } from './hooks';
import { defaultOnLoad, findCheckedValue, systemColumns } from './helpers';
import { RegistryContext } from '../../store';
import { InventoryTable } from '@redhat-cloud-services/frontend-components/Inventory';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import { SystemColumn } from './types';
import { useDispatch } from 'react-redux';
import { useRbac } from '../../Helpers/Hooks';
import { PERMISSIONS } from '../../Helpers/constants';
import { NotAuthorized } from '@redhat-cloud-services/frontend-components/NotAuthorized';
import { fetchSystems } from '../../api';

import './InventoryPage.scss';

export const InventoryPage = ({ selectedIds, setSelectedIds }) => {
  const chrome = useChrome();
  const dispatch = useDispatch();
  const inventory = useRef(null);
  const { getRegistry } = useContext(RegistryContext);

  const [activeFiltersString, setActiveFiltersString] = useState('');
  const [isBulkLoading, setBulkLoading] = useState(false);
  const [items, setItems] = useState([]);
  const [total, setTotal] = useState(0 as number);
  const [[canReadHostsInventory], isLoadingInventory] = useRbac([PERMISSIONS.readHosts], 'inventory');

  useEffect(() => {
    dispatch({ type: 'INVENTORY_INIT' });
    dispatch({ type: 'RESET_PAGE' });
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    inventory?.current?.onRefreshData();
  }, [dispatch]);

  useEffect(() => {
    dispatch({
      type: 'SELECT_ENTITIES',
      payload: {
        selected: selectedIds,
      },
    });
  }, [selectedIds]);

  const onComplete = (result, filtersString: string) => {
    setTotal(result.total);
    setItems(result.results);
    setActiveFiltersString(filtersString);
  };

  const getEntities = useGetEntities(onComplete, {
    selectedIds,
  });

  const mergedColumns = (defaultColumns) =>
    systemColumns(chrome?.isBeta?.()).map((column: SystemColumn) => {
      const defaultColumn = defaultColumns.find((defaultCol) => defaultCol.key === column.key);
      return {
        ...defaultColumn,
        ...column,
        props: {
          ...defaultColumn?.props,
          ...column?.props,
        },
      };
    });

  const selectIds = (_event, _isSelected, _index, entity) => {
    const newSelectedIds: string[] = [...selectedIds];

    !newSelectedIds.includes(entity.id) ? newSelectedIds.push(entity.id) : newSelectedIds.splice(newSelectedIds.indexOf(entity.id), 1);

    setSelectedIds(newSelectedIds);
  };

  const bulkSelectIds = async (type: string, options?) => {
    const newSelectedIds = [...selectedIds];

    setBulkLoading(true);
    switch (type) {
      case 'none': {
        setSelectedIds([]);
        break;
      }

      case 'page': {
        options.items.forEach((item) => {
          if (!newSelectedIds.includes(item.id)) {
            newSelectedIds.push(item.id);
          }
        });

        setSelectedIds(newSelectedIds);
        break;
      }

      case 'page-none': {
        options.items.forEach((item) => {
          if (newSelectedIds.includes(item.id)) {
            newSelectedIds.splice(newSelectedIds.indexOf(item.id), 1);
          }
        });

        setSelectedIds(newSelectedIds);
        break;
      }

      case 'all': {
        const ids = [] as string[];
        const perPage = 100;
        const totalPages = Math.ceil(total / perPage);
        for (let page = 1; page <= totalPages; page++) {
          const { results: pageResults } = await fetchSystems(`?per_page=100&page=${page}${activeFiltersString}`);
          ids.push(...pageResults.map(({ id }) => id));
        }
        setSelectedIds(ids);
        break;
      }
    }
    setBulkLoading(false);
  };

  return isLoadingInventory ? (
    <Bullseye>
      <Spinner />
    </Bullseye>
  ) : canReadHostsInventory ? (
    <InventoryTable
      isFullView
      autoRefresh
      showTags
      hideFilters={{
        all: true,
        name: false,
        operatingSystem: false,
      }}
      columns={mergedColumns}
      ref={inventory}
      fallback={
        <Bullseye>
          <Spinner />
        </Bullseye>
      }
      onLoad={defaultOnLoad(getRegistry)}
      getEntities={getEntities}
      tableProps={{
        isStickyHeader: true,
        isStriped: true,
        canSelectAll: false,
        onSelect: items.length ? selectIds : null,
      }}
      bulkSelect={{
        id: 'systems-bulk-select',
        isDisabled: !total,
        items: [
          {
            title: `Select none (0)`,
            onClick: () => {
              bulkSelectIds('none');
            },
          },
          {
            title: `Select page (${items?.length || 0})`,
            onClick: () => {
              bulkSelectIds('page', { items: items });
            },
          },
          {
            title: `Select all (${total || 0})`,
            onClick: () => {
              bulkSelectIds('all', { total: total });
            },
          },
        ],
        onSelect: () => {
          const allOnPageSelected = findCheckedValue(items, selectedIds);
          if (allOnPageSelected) {
            bulkSelectIds('page-none', { items: items });
          } else {
            bulkSelectIds('page', { items: items });
          }
        },
        checked: items && selectedIds ? findCheckedValue(items, selectedIds) : null,
        toggleProps: {
          'data-ouia-component-type': 'bulk-select-toggle-button',
          children: isBulkLoading
            ? [
                <React.Fragment key="sd">
                  <Spinner size="sm" />
                  {` ${selectedIds.length} selected `}
                </React.Fragment>,
              ]
            : `${selectedIds.length} selected `,
        },
      }}
    />
  ) : (
    <NotAuthorized serviceName="Upgrades" />
  );
};

InventoryPage.propTypes = {
  selectedIds: propTypes.arrayOf(propTypes.string),
  setSelectedIds: propTypes.func,
};

export default InventoryPage;
