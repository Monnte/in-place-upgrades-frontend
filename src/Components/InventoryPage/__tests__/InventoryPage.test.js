import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import InventoryPage from '../InventoryPage';
import { useGetEntities } from '../hooks';
import { findCheckedValue } from '../Helpers';
import { useRbac } from '../../../Helpers/hooks';

const systemsMock = {
  total: 2,
  count: 2,
  page: 1,
  per_page: 10,
  results: [
    {
      display_name: 'rhiqe.8d8fe668-c2ee-4034-b174-2af31020402c.email-38.alvarez.com',
      id: '1d2a2d8b-8bb1-48f3-a367-c8ba2f4d8fc3',
      system_profile: {
        operating_system: {
          major: 8,
          minor: 6,
          name: 'RHEL',
        },
      },
      selected: true,
      updated: '2022-04-20T12:49:29.389485Z',
    },
    {
      display_name: 'rhiqe.laptop-44.sanchez-mann.biz',
      id: 'aa28f81c-e7f0-4ee4-af46-0352b0fb50f3',
      system_profile: {
        operating_system: {
          major: 8,
          minor: 8,
          name: 'RHEL',
        },
      },
      selected: false,
      updated: '2022-05-04T07:08:37.176240Z',
    },
  ],
};

const systemTags = {
  total: 2,
  count: 2,
  page: 1,
  per_page: 10,
  results: {
    '1d2a2d8b-8bb1-48f3-a367-c8ba2f4d8fc3': [
      {
        namespace: 'insights-client',
        key: 'application',
        value: 'TEST',
      },
      {
        namespace: 'insights-client',
        key: 'cloud',
        value: 'POG',
      },
    ],
  },
};

jest.mock('../../../api', () => {
  return {
    inventoryFetchSystems: jest.fn(() => Promise.resolve(systemsMock)),
    inventoryFetchSystemsTags: jest.fn(() => Promise.resolve(systemTags)),
  };
});

jest.mock('../../../Helpers/hooks');

describe('InventoryTable', () => {
  let mockStore = configureStore();
  let props;
  const setSelectedIds = jest.fn();

  beforeEach(() => {
    useRbac.mockImplementation(() => [[true], false]);
  });

  it('should render correctly', () => {
    const store = mockStore(props);
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <InventoryPage selectedIds={[]} setSelectedIds={setSelectedIds} />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly without permissions', () => {
    useRbac.mockImplementation(() => [[false], false]);
    const store = mockStore(props);
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <InventoryPage selectedIds={[]} setSelectedIds={setSelectedIds} />
        </Router>
      </Provider>
    );
    expect(asFragment().textContent).toMatch(/You do not have access to Upgrades/);
  });

  it('useGetEntities hook should work', async () => {
    const onComplete = jest.fn();
    const getEntities = useGetEntities(onComplete, { selectedIds: ['1d2a2d8b-8bb1-48f3-a367-c8ba2f4d8fc3'] });
    const _items = jest.fn();
    const config = {
      page: 1,
      per_page: 10,
      orderBy: 'updated',
      orderDirection: 'DESC',
      filters: {
        staleFilter: undefined,
        lastSeenFilter: undefined,
        registeredWithFilter: undefined,
        osFilter: ['8.0', '8.6'],
        hostGroupFilter: undefined,
        hostnameOrId: 'blob',
      },
    };
    const entities = await getEntities(_items, config);
    const resultSystems = systemsMock.results.map((system) => {
      return {
        ...system,
        tags: systemTags.results[system.id] || [],
      };
    });

    expect(entities.results).toEqual(resultSystems);
    expect(onComplete).toHaveBeenCalled();

    // check get entities without selectedIds
    const getEntitiesWithoutSelectedIds = useGetEntities(onComplete, {});
    const entitiesWithoutSelectedIds = await getEntitiesWithoutSelectedIds(_items, config);
    const systemsMockWithoutSelected = systemsMock.results.map((system) => {
      return {
        ...system,
        selected: false,
        tags: systemTags.results[system.id] || [],
      };
    });
    expect(entitiesWithoutSelectedIds.results).toEqual(systemsMockWithoutSelected);
    expect(onComplete).toHaveBeenCalled();
  });

  it('findCheckedValue helper should work', () => {
    const checkedValue = findCheckedValue(systemsMock.results, []);
    expect(checkedValue).toEqual(false);

    const checkedValue2 = findCheckedValue(systemsMock.results, ['1d2a2d8b-8bb1-48f3-a367-c8ba2f4d8fc3']);
    expect(checkedValue2).toEqual(null);

    const checkedValue3 = findCheckedValue(systemsMock.results, ['1d2a2d8b-8bb1-48f3-a367-c8ba2f4d8fc3', 'aa28f81c-e7f0-4ee4-af46-0352b0fb50f3']);
    expect(checkedValue3).toEqual(true);
  });
});
