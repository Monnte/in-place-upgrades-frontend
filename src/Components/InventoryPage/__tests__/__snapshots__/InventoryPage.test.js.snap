// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryTable should render correctly 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Inventory mock
    </h1>
  </div>
</DocumentFragment>
`;
